pkg/plugin/v3/init.go

```diff
@@ -0,0 +1,165 @@
+- a Patch file for customizing image for manager manifests
+- a Patch file for enabling prometheus metrics
+- a cmd/manager/main.go to run
+
+project will prompt the user to run 'dep ensure' after writing the project files.
+`
+	ctx.Examples = fmt.Sprintf(`  # Scaffold a project using the apache2 license with "The Kubernetes authors" as owners
+  %s init --project-version=2 --domain example.org --license apache2 --owner "The Kubernetes authors"
+`,
+		ctx.CommandName)
+
+	p.commandName = ctx.CommandName
+}
+
+func (p *initPlugin) BindFlags(fs *pflag.FlagSet) {
+	fs.BoolVar(&p.skipGoVersionCheck, "skip-go-version-check",
+		false, "if specified, skip checking the Go version")
+
+	// dependency args
+	fs.BoolVar(&p.fetchDeps, "fetch-deps", true, "ensure dependencies are downloaded")
+
+	// boilerplate args
+	fs.StringVar(&p.license, "license", "apache2",
+		"license to use to boilerplate, may be one of 'apache2', 'none'")
+	fs.StringVar(&p.owner, "owner", "", "owner to add to the copyright")
+
+	// project args
+	fs.StringVar(&p.config.Repo, "repo", "", "name to use for go module (e.g., github.com/user/repo), "+
+		"defaults to the go package of the current working directory.")
+	fs.StringVar(&p.config.Domain, "domain", "my.domain", "domain for groups")
+}
+
+func (p *initPlugin) InjectConfig(c *config.Config) {
+	c.Layout = plugin.KeyFor(Plugin{})
+	p.config = c
+}
+
+func (p *initPlugin) Run() error {
+	return cmdutil.Run(p)
+}
+
+func (p *initPlugin) Validate() error {
+	// Requires go1.11+
+	if !p.skipGoVersionCheck {
+		if err := util.ValidateGoVersion(); err != nil {
+			return err
+		}
+	}
+
+	// Check if the project name is a valid namespace according to k8s
+	dir, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("error to get the current path: %v", err)
+	}
+	projectName := filepath.Base(dir)
+	if err := validation.IsDNS1123Label(strings.ToLower(projectName)); err != nil {
+		return fmt.Errorf("project name (%s) is invalid: %v", projectName, err)
+	}
+
+	// Try to guess repository if flag is not set.
+	if p.config.Repo == "" {
+		repoPath, err := util.FindCurrentRepo()
+		if err != nil {
+			return fmt.Errorf("error finding current repository: %v", err)
+		}
+		p.config.Repo = repoPath
+	}
+
+	return nil
+}
+
+func (p *initPlugin) GetScaffolder() (scaffold.Scaffolder, error) {
+	return scaffolds.NewInitScaffolder(p.config, p.license, p.owner), nil
+}
+
+func (p *initPlugin) PostScaffold() error {
+	if !p.fetchDeps {
+		fmt.Println("Skipping fetching dependencies.")
+		return nil
+	}
+
+	// Ensure that we are pinning controller-runtime version
+	// xref: https://github.com/kubernetes-sigs/kubebuilder/issues/997
+	err := util.RunCmd("Get controller runtime", "go", "get",
+		"sigs.k8s.io/controller-runtime@"+scaffolds.ControllerRuntimeVersion)
+	if err != nil {
+		return err
+	}
+
+	err = util.RunCmd("Update go.mod", "go", "mod", "tidy")
+	if err != nil {
+		return err
+	}
+
+	err = util.RunCmd("Running make", "make")
```

Comment:
shouldn't this be conditional on the make flag defined in the other file?

Reply:
Currently there's no `init --make` flag unlike `create api`. Perhaps we should add one.

Reply:
Should we not here just replicate so the code to v3? IMO we should not change any impl in this pr just add the v3-alpha plugin and then we can check each case/scenario in new pr's.

Reply:
This won't be changed here, just added a TODO.